File: answers/leetcode/290. Word Pattern/Solution.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split(' ')
        if len(pattern) != len(words):
            return False
        letter_map = dict()
        word_map = dict()
        for i in range(len(pattern)):
            if pattern[i] not in letter_map and words[i] not in word_map:
                letter_map[pattern[i]] = words[i]
                word_map[words[i]] = pattern[i]
            elif pattern[i] not in letter_map or words[i] not in word_map or letter_map[pattern[i]] != words[i]:
                return False
        return True

File: answers/leetcode/290. Word Pattern/Solution.cpp
//@type  Hash Table
//@result 26 / 26 test cases passed. Status: Accepted Runtime: 0 ms Submitted: 0 minutes ago

class Solution {
public:
    bool wordPattern(string pattern, string str) {
		istringstream iss(str);
		vector<string> tokens{istream_iterator<string>{iss}, istream_iterator<string>{}};
		if (pattern.length() != tokens.size()) {
			return false;
		}
		set<string> word_set;
		map<char, string> word_map;
		for (size_t i = 0; i < pattern.length(); ++ i) {
			if (word_map.end() == word_map.find(pattern[i])) {
				if (word_set.end() != word_set.find(tokens[i])) {
					return false;
				}
				word_set.insert(tokens[i]);
				word_map.insert(make_pair(pattern[i], tokens[i]));
			}
			else if (word_map[pattern[i]] != tokens[i]) {
				return false;
			}
		}
		return true;
    }
};